keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,585 @@ yet but we think may be interesting to consider in the future.
+rejection will cause the client to slow down.  Lacking that
+confidence, we choose to reject the youngest rather than the oldest
+request of the queue, so that an investment in holding a request in a
+queue has a chance of eventually getting useful work done.
+
+### Dispatching
+
+Requests of an exempt priority are never held up in a queue; they are
+always dispatched immediately.  Following is how the other requests
+are dispatched at a given apiserver.
+
+The concurrency limit of an apiserver is divided among the non-exempt
+priority levels in proportion to their assured concurrency shares.
+This produces the assured concurrency value (ACV) for each non-exempt
+priority level:
+
+```
+ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+```
+
+where SCL is the apiserver's concurrency limit and ACS(l) is the
+AssuredConcurrencyShares for priority level l.
+
+Dispatching is done independently for each priority level.  Whenever
+(1) a non-exempt priority level's number of running requests is below
+the level's assured concurrency value and (2) that priority level has
+a non-empty queue, it is time to dispatch another request for service.
+The Fair Queuing for Server Requests algorithm below is used to pick a
+non-empty queue at that priority level.  Then the request at the head
+of that queue is dispatched.
+
+
+#### Fair Queuing for Server Requests
+
+This is based on fair queuing but is modified to deal with serving
+requests in an apiserver instead of transmitting packets in a router.
+You can find the original fair queuing paper at
+[ACM](https://dl.acm.org/citation.cfm?doid=75247.75248) or
+[MIT](http://people.csail.mit.edu/imcgraw/links/research/pubs/networks/WFQ.pdf),
+and an
+[implementation outline at Wikipedia](https://en.wikipedia.org/wiki/Fair_queuing).
+Our problem differs from the normal fair queuing problem in three
+ways.  One is that we are dispatching requests to be served rather
+than packets to be transmitted.  Another difference is that multiple
+requests may be served at once.  The third difference is that the
+actual service time (i.e., duration) is not known until a request is
+done being served.  The first two differences can easily be handled by
+straightforward adaptation of the concept called "R(t)" in the
+original paper and "virtual time" in the implementation outline.  In
+that implementation outline, the notation `now()` is used to mean
+reading the _virtual_ clock.  In the original paper’s terms, "R(t)" is
+the number of "rounds" that have been completed at real time t, where
+a round consists of virtually transmitting one bit from every
+non-empty queue in the router (regardless of which queue holds the
+packet that is really being transmitted at the moment); in this
+conception, a packet is considered to be "in" its queue until the
+packet’s transmission is finished.  For our problem, we can define a
+round to be giving one nanosecond of CPU to every non-empty queue in
```

Comment:
In a single round, is only one packet dequeud/sent?

Reply:
Remember that the rounds describe a fantasy, and that fantasy is used to guide what really happens.  In the Demers paper a round consists of transmitting one bit from each queue; the actual transmission does not go that way.  The actual packet transmissions are ordered according to completion times of packets in these fantasy rounds.